File: installers/dev_tools.py
# SPDX-FileCopyrightText: Copyright (C) 2025 Akshat Kotpalliwar (alias IntegerAlex) <inquiry.akshatkotpalliwar@gmail.com>
# SPDX-License-Identifier: GPL-3.0-only
"""Developer tools installer."""

import shutil

from utils.base import BaseInstaller
from utils.types import InstallerResult


class DevToolsInstaller(BaseInstaller):
    """Install developer tools."""
    
    def is_installed(self) -> bool:
        # Check if key developer tools are installed
        # We check for a few key tools to determine if dev tools are installed
        key_tools = ["gcc", "make", "cmake", "pkg-config"]
        installed_count = sum(1 for tool in key_tools if shutil.which(tool))
        # If at least 3 out of 4 key tools are installed, consider dev tools installed
        return installed_count >= 3
    
    def install(self) -> InstallerResult:
        try:
            self.console.print("[blue]ℹ[/blue] Installing developer tools")
            
            packages = [
                "openssl",
                "clangd",
                "build-essential",
                "pkg-config",
                "cmake",
                "git"
            ]
            
            self.runner.run(
                ["apt", "install", "-y"] + packages,
                sudo=True,
                description="Installing developer tools"
            )
            
            return InstallerResult(True, "Developer tools installed successfully")
        except Exception as e:
            return InstallerResult(False, "Failed to install developer tools", str(e))

File: src/installers/dev-tools.ts
import type { CommandRunner } from "../utils/command-runner.js";
import type { SetupLogger } from "../utils/logger.js";
import type { InstallerResult } from "../types.js";
import { blue } from "../utils/ansi.js";

export function createDevToolsInstaller(runner: CommandRunner, logger: SetupLogger) {
  function checkInstalled(): boolean {
    const keyTools = ["gcc", "make", "cmake", "pkg-config"];
    let count = 0;
    for (const tool of keyTools) {
      try {
        const r = runner.run(["which", tool], { check: false, capture: true });
        if ((r.status ?? 1) === 0) count++;
      } catch {}
    }
    return count >= 3;
  }

  async function doInstall(): Promise<InstallerResult> {
    try {
      process.stdout.write(`${blue("\u2139")} Installing developer tools\n`);
      const packages = ["openssl", "clangd", "build-essential", "pkg-config", "cmake", "git"];
      runner.run(
        ["apt", "install", "-y", ...packages],
        { sudo: true, description: "Installing developer tools" },
      );
      return { success: true, message: "Developer tools installed successfully" };
    } catch (e) {
      return { success: false, message: "Failed to install developer tools", error: String(e) };
    }
  }

  return { is_installed: checkInstalled, install: doInstall };
}
